File: internal/tool/plan.go
package tool

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// What a task plan is, and what a surface has to be able to do with one.
//
// The vocabulary lives here, next to the Tool interface, for the same reason the
// ask_user vocabulary does (see ask.go): three packages have to agree on it and
// none of them owns the others. A tool writes a plan (internal/tool/builtin), a
// turn carries the planner on its context (internal/server), and the console
// renders and stores the same shape (internal/server, web/src/plan.js). Putting
// the types in any one of them would make the other two depend on it for no
// reason.
//
// Nothing here knows about SQLite, SSE or the browser: a Planner decides where a
// plan lives, who may change it, and what happens when the model asks for
// something impossible.

// TaskStatus is where one task of a plan stands.
//
// The five values are the whole vocabulary: a board renders them as three groups
// (执行中 / 待执行 / 已完成), and "failed" stays in 待执行 because a failed step is
// still a step to do — it is unfinished work with a reason attached, not a
// finished one.
type TaskStatus string

const (
	// TaskPending is not started.
	TaskPending TaskStatus = "pending"
	// TaskInProgress is being worked on right now. At most one task should be in
	// this state: it is what the board's 执行中 group and the model's own "where
	// am I" both read.
	TaskInProgress TaskStatus = "in_progress"
	// TaskDone is finished.
	TaskDone TaskStatus = "done"
	// TaskFailed was attempted and did not work.
	TaskFailed TaskStatus = "failed"
	// TaskSkipped was deliberately not done (no longer needed, superseded).
	TaskSkipped TaskStatus = "skipped"
)

// ValidTaskStatus reports whether s is one of the five statuses.
func ValidTaskStatus(s TaskStatus) bool {
	switch s {
	case TaskPending, TaskInProgress, TaskDone, TaskFailed, TaskSkipped:
		return true
	}
	return false
}

// TaskStatusList renders the vocabulary for a model-facing error message, so a
// model that invented a status is told the real ones instead of just "no".
func TaskStatusList() string {
	return string(TaskPending) + " / " + string(TaskInProgress) + " / " +
		string(TaskDone) + " / " + string(TaskFailed) + " / " + string(TaskSkipped)
}

// Task is one item of a plan. The JSON tags are the wire contract with the
// console (web/src/plan.js), so they are snake_case like the rest of the API.
type Task struct {
	// ID is stable for the life of the plan and is how the model addresses the
	// task when it updates it. The surface assigns it (a model that invents its
	// own ids produces duplicates across calls), so a Task coming from a tool
	// call may leave it empty.
	ID     string     `json:"id"`
	Title  string     `json:"title"`
	Status TaskStatus `json:"status"`
	// Note is the optional one-line remark: what was found, why it failed, what
	// is left. It is what makes a finished task informative a turn later.
	Note string `json:"note,omitempty"`
}

// Plan is the task list for one conversation.
//
// It is deliberately flat and small: it is read by a model deciding what to do
// next and by a person glancing above the input box, and neither of them is
// served by a dependency graph.
type Plan struct {
	// Goal is what the whole plan is for, in the user's terms. It is what makes a
	// resumed turn possible at all — a task list with no subject is a list of
	// chores.
	Goal string `json:"goal"`
	// Tasks are the items, in the order the model intends to do them.
	Tasks []Task `json:"tasks"`
	// Revision increments on every change. The console uses it to tell a newer
	// plan from an older one; nothing else depends on it.
	Revision int `json:"revision"`
	// UpdatedAt is when the surface last changed this plan, in its own clock.
	UpdatedAt time.Time `json:"updated_at"`
}

// Empty reports whether the plan carries no tasks. An empty plan is treated as
// "no plan" everywhere: the board does not render, and 继续执行 has nothing to
// resume.
func (p Plan) Empty() bool { return len(p.Tasks) == 0 }

// Count returns how many tasks carry the given status.
func (p Plan) Count(status TaskStatus) int {
	n := 0
	for _, t := range p.Tasks {
		if t.Status == status {
			n++
		}
	}
	return n
}

// Progress returns how many tasks are done and how many there are.
func (p Plan) Progress() (done, total int) {
	return p.Count(TaskDone), len(p.Tasks)
}

// Unfinished returns how many tasks are not done. Skipped tasks count as
// finished: someone decided they were not needed, which is a decision, not a
// leftover.
func (p Plan) Unfinished() int {
	n := 0
	for _, t := range p.Tasks {
		if t.Status != TaskDone && t.Status != TaskSkipped {
			n++
		}
	}
	return n
}

// Task returns the task with this id and its position, or false when there is
// none. IDs are matched exactly: a model that passes "1" for task "t1" is
// corrected with the real ids rather than guessed at, because a plan where the
// wrong task is marked done is worse than a failed tool call.
func (p Plan) Task(id string) (Task, int, bool) {
	for i, t := range p.Tasks {
		if t.ID == id {
			return t, i, true
		}
	}
	return Task{}, -1, false
}

// IDs renders every task id, for an error message that lets the model fix its
// own call.
func (p Plan) IDs() string {
	if len(p.Tasks) == 0 {
		return "（计划为空）"
	}
	out := make([]string, 0, len(p.Tasks))
	for _, t := range p.Tasks {
		out = append(out, t.ID)
	}
	return strings.Join(out, ", ")
}

// Summary is one line about the plan, for a board header and for the model.
func (p Plan) Summary() string {
	if p.Empty() {
		return "计划为空"
	}
	done, total := p.Progress()
	out := fmt.Sprintf("已完成 %d/%d", done, total)
	var doing []string
	for _, t := range p.Tasks {
		if t.Status == TaskInProgress {
			doing = append(doing, t.Title)
		}
	}
	switch {
	case len(doing) > 0:
		out += " · 进行中：" + strings.Join(doing, "、")
	case p.Unfinished() == 0:
		out += " · 全部完成"
	default:
		out += fmt.Sprintf(" · 待执行 %d 项", p.Unfinished())
	}
	return out
}

// Render is the plan as a checklist, which is what a model reads back.
//
// The markers are text rather than the raw status values on purpose: a model that
// sees "[x]" and "[ ]" recognises the convention, while "status=done" makes it
// guess whether the rest of the line still matters.
func (p Plan) Render() string {
	var b strings.Builder
	if strings.TrimSpace(p.Goal) != "" {
		b.WriteString("目标：" + p.Goal + "\n")
	}
	if p.Empty() {
		b.WriteString("（还没有任务）")
		return b.String()
	}
	b.WriteString("进度：" + p.Summary() + "\n")
	for i, t := range p.Tasks {
		b.WriteString(strconv.Itoa(i+1) + ". " + taskMarker(t.Status) + " " + t.ID + " " + t.Title)
		if strings.TrimSpace(t.Note) != "" {
			b.WriteString(" — " + t.Note)
		}
		b.WriteString("\n")
	}
	return strings.TrimRight(b.String(), "\n")
}

// taskMarker renders one status as the checklist glyph above.
func taskMarker(s TaskStatus) string {
	switch s {
	case TaskDone:
		return "[x]"
	case TaskInProgress:
		return "[>]"
	case TaskFailed:
		return "[!]"
	case TaskSkipped:
		return "[-]"
	default:
		return "[ ]"
	}
}

// NextTaskID returns an id that no task in tasks is using, in the "t<n>" form the
// surface assigns.
//
// It walks past the highest numeric suffix rather than counting tasks, so a plan
// that lost a task to a validation error — or one created in several calls —
// cannot hand out the same id twice.
func NextTaskID(tasks []Task) string {
	highest := 0
	for _, t := range tasks {
		n, ok := taskIDSuffix(t.ID)
		if ok && n > highest {
			highest = n
		}
	}
	return "t" + strconv.Itoa(highest+1)
}

// taskIDSuffix parses the number out of a "t<n>" id.
func taskIDSuffix(id string) (int, bool) {
	if len(id) < 2 || id[0] != 't' {
		return 0, false
	}
	n, err := strconv.Atoi(id[1:])
	if err != nil || n <= 0 {
		return 0, false
	}
	return n, true
}

// TaskPatch is one change to one task. Every field is optional, and the zero
// value of each means "leave it alone" — which is why Note is a pointer: "no
// note" and "remove the note" are different requests, and a planner that
// conflated them would silently erase the reason a task failed.
type TaskPatch struct {
	Title  string
	Status TaskStatus
	Note   *string
}

// Empty reports whether the patch would change nothing, which the tool refuses
// rather than writing a no-op that looks like work.
func (p TaskPatch) Empty() bool {
	return strings.TrimSpace(p.Title) == "" && p.Status == "" && p.Note == nil
}

// Planner is a surface's plan store for one conversation.
//
// It is read from the turn's context (PlannerFrom) rather than captured by the
// tools, for the same reason the Asker is: one tool set is shared by every
// conversation of a surface, and each turn belongs to its own.
//
// Implementations own the limits — how many tasks a plan may hold, what an id
// looks like, whether a change is persisted at all — because those are
// deployment decisions. The tools own the parts the model has to fix itself: an
// empty title, a status that is not in the vocabulary, a task id that does not
// exist.
type Planner interface {
	// Current returns the plan as it stands, and whether there is one.
	Current(ctx context.Context) (Plan, bool, error)
	// Replace creates or overwrites the plan wholesale. It is what plan_create
	// calls, and it is the only way a plan comes into existence.
	Replace(ctx context.Context, plan Plan) (Plan, error)
	// Append adds tasks to the end of the current plan.
	Append(ctx context.Context, tasks []Task) (Plan, error)
	// Update changes one task, addressed by id.
	Update(ctx context.Context, taskID string, patch TaskPatch) (Plan, error)
}

// plannerKey is the context key carrying a turn's Planner. Unexported so the
// only way to install one is WithPlanner.
type plannerKey struct{}

// WithPlanner returns ctx carrying p, the plan store for this turn.
//
// A nil Planner returns ctx unchanged: "this surface has no plans" is the
// absence of the value, not a value every reader has to nil-check.
func WithPlanner(ctx context.Context, p Planner) context.Context {
	if p == nil || ctx == nil {
		return ctx
	}
	return context.WithValue(ctx, plannerKey{}, p)
}

// PlannerFrom returns the Planner installed on ctx, and whether there is one.
//
// A tool that gets false must say so instead of pretending to record a plan: the
// tools are only registered on surfaces that install a planner, so this is the
// belt to that braces — a surface that adds the tools without the store gets a
// readable refusal rather than a plan nobody kept.
func PlannerFrom(ctx context.Context) (Planner, bool) {
	if ctx == nil {
		return nil, false
	}
	p, ok := ctx.Value(plannerKey{}).(Planner)
	if !ok || p == nil {
		return nil, false
	}
	return p, true
}

File: web/src/plan.js
// 任务看板的数据模型（纯函数，不依赖 Vue）。
//
// 服务端工具（`tool.Plan` / `tool.Task`，snake_case）和这里是同一份数据的两种
// 写法，本文件只做三件事：把 wire 形状归一成能直接渲染的形状、按状态分组、
// 算出一行摘要。
//
// 它单独成一个模块（而不是写进组件里）有两个原因：一是这些规则值得在没有浏览器
// 的时候被断言（见 ssr-probe/run.mjs 与 scripts/check-*.mjs 的同类做法）；二是
// 看板因此只剩下循环和样式，判断"哪一项算待执行"不会散落在模板条件里。
//
// 每一份计划都是模型写给人看的：它可能不完整（缺字段、状态是它自己编的），所以
// 归一化的原则是**容忍**——能显示的就显示，实在没内容的才丢，而不是因为一处脏
// 数据让整块看板消失。

/** 计划允许的五个状态，与服务端 `tool.Task` 的枚举一致。 */
export const PLAN_STATUSES = ['pending', 'in_progress', 'done', 'failed', 'skipped']

/**
 * 三个分组的标题。
 *
 * 写在这里而不是组件里：分组的含义（failed 为什么在待执行、skipped 算什么）是
 * 模型的一部分，标题只是它的一个说法——换说法不该要求改分组规则。
 */
export const PLAN_GROUP_LABELS = { doing: '执行中', todo: '待执行', done: '已完成' }

/**
 * 取值成一个字符串。
 *
 * 数字也算有值（`title: 30` 是一句合法的任务标题），对象 / 数组 / null 不算：
 * 它们拼进模板只会变成 "[object Object]"。
 */
function text(value) {
  if (typeof value === 'string') return value.trim()
  if (typeof value === 'number' && Number.isFinite(value)) return String(value)
  return ''
}

/** 计数字段：非有限数、负数、字符串里的数字都收敛成一个非负整数。 */
function count(value) {
  const n = Number(value)
  return Number.isFinite(n) && n > 0 ? Math.floor(n) : 0
}

/**
 * 一个没被占用的补号 id。
 *
 * 服务端会分配 `t1`、`t2`…，但一份本地造的或半截的计划可能只有标题。补号用
 * **它在原数组里的序号**，这样同一份 JSON 每次归一化得到的 id 都一样（否则看板
 * 的 `:key` 每帧都在变，列表会被重建）；万一和别的任务撞了，就往后找一个空的。
 */
function freeId(base, used) {
  if (!used.has(base)) return base
  let n = 2
  while (used.has(`${base}-${n}`)) n += 1
  return `${base}-${n}`
}

/**
 * normalizePlan 把 wire 形状（或任何像计划的东西）读成看板渲染的形状。
 *
 * 返回 `null` 表示"没有计划"，看板据此**整个不渲染**——"没计划"和"计划里没有
 * 任务"对界面是同一件事（没有可显示的内容），服务端也用 `null` 表示空计划。
 */
export function normalizePlan(raw) {
  if (!raw || typeof raw !== 'object') return null

  const tasks = []
  const used = new Set()
  const list = Array.isArray(raw.tasks) ? raw.tasks : []
  list.forEach((item, index) => {
    const source = item && typeof item === 'object' ? item : {}
    // 没有标题的任务没有可渲染的内容，丢掉它而不是画一行空白：看板上"还剩几件
    // 事"这个数字是给人做判断用的，把空行算进去会让它不可信。
    const title = text(source.title)
    if (title === '') return

    let id = text(source.id)
    // 服务端给的 id 原样保留（`plan_update` 靠它定位，改名等于撒谎）；只有真的
    // 没有 id 时才补号。
    if (id === '') id = freeId(`t${index + 1}`, used)
    used.add(id)

    tasks.push({
      id,
      title,
      // 状态只认五个枚举值；别的（拼错的、服务端将来新增的）当待执行——待执行
      // 是最保守的一档：它不会让一项工作从看板上消失。
      status: PLAN_STATUSES.includes(source.status) ? source.status : 'pending',
      note: text(source.note),
    })
  })

  if (tasks.length === 0) return null

  return {
    goal: text(raw.goal),
    revision: count(raw.revision),
    // 服务端每次变更 +1，客户端只读。看板今天不显示这一对（进度与分组已经说清了
    // "到哪了"），但归一化里仍然留着：把它丢掉等于替调用方决定它没用。
    updatedAt: text(raw.updated_at) || text(raw.updatedAt),
    tasks,
  }
}

/**
 * planGroups 把任务分到看板的三个组里。
 *
 * `failed` 和 `skipped` 都留在「待执行」而不是自成一组：对"还剩什么"这个问题，
 * 它们都还是没做完的事。`failed` 排在待执行的最前面，因为它是这一组里唯一需要
 * 人看一眼的（要么重跑，要么换个做法），而 `pending` 只是排队。
 */
export function planGroups(raw) {
  const groups = { doing: [], todo: [], done: [] }
  const plan = normalizePlan(raw)
  if (!plan) return groups

  for (const task of plan.tasks) {
    if (task.status === 'in_progress') groups.doing.push(task)
    else if (task.status === 'done') groups.done.push(task)
    else groups.todo.push(task)
  }
  groups.todo = [
    ...groups.todo.filter((task) => task.status === 'failed'),
    ...groups.todo.filter((task) => task.status !== 'failed'),
  ]
  return groups
}

/**
 * planProgress 是头部那三个数字：已完成数、总数、百分比。
 *
 * 百分比取整（0..100）：它旁边的 `n/m` 才是精确值，百分比只是同一条信息的
 * 另一种读法，"5/6 → 83%" 比 "83.33%" 好读。
 */
export function planProgress(raw) {
  const plan = normalizePlan(raw)
  const total = plan ? plan.tasks.length : 0
  const done = plan ? plan.tasks.filter((task) => task.status === 'done').length : 0
  const percent = total === 0 ? 0 : Math.min(100, Math.max(0, Math.round((done / total) * 100)))
  return { done, total, percent }
}

/**
 * planResumable：计划里还有真正没做完的任务 → 可以「继续执行」。
 *
 * `skipped` 不算没做完：跳过是"这一项不需要做了"的决定，是一个结论而不是遗留
 * 的工作。服务端 tool.Plan.Unfinished() 的算法与这里一致，两边必须同口径——否则
 * 看板会在服务端认为无事可续的时候亮着按钮，点下去只会收到 400。
 *
 * `failed` 算：它恰好是要续跑的理由。
 */
export function planResumable(raw) {
  const plan = normalizePlan(raw)
  return Boolean(plan) && plan.tasks.some((task) => task.status !== 'done' && task.status !== 'skipped')
}

/**
 * planHeadline 是一行摘要，给看板的折叠态和 title 用。
 *
 * 先说总数（`已完成 2/5`），再说"现在在做什么"或"下一步是什么"：折叠之后这一行
 * 就是全部信息，只写 2/5 会让人不知道差在哪一步。没有计划时返回空串，调用方
 * 不必再判空。
 */
export function planHeadline(raw) {
  const plan = normalizePlan(raw)
  if (!plan) return ''

  const { done, total } = planProgress(plan)
  const groups = planGroups(plan)
  const parts = [`已完成 ${done}/${total}`]

  if (groups.doing.length) {
    const running = groups.doing[0].title
    parts.push(`进行中：${running}${groups.doing.length > 1 ? ` 等 ${groups.doing.length} 项` : ''}`)
  } else if (groups.todo.length) {
    const next = groups.todo[0]
    parts.push(`${next.status === 'failed' ? '失败' : '下一步'}：${next.title}`)
  } else {
    parts.push('全部完成')
  }

  return parts.join(' · ')
}
